File: iosApp/iosApp/AppState.swift
import Foundation
import Observation
import SwiftUI

/// Root observable state for the entire app.
///
/// `@Observable` (Swift 5.9+) re-renders only views that read each property —
/// no manual `@Published` boilerplate. `Set<String>` of seen signal IDs lets
/// us detect newly-arrived signals on each run and trigger mock pushes.
/// Backend URL — hardcoded, served from the Pi via Tailscale Funnel.
enum BackendURL {
    static let url = "https://raspberrypi.taile76757.ts.net"
}

@Observable
@MainActor
final class AppState {
    private(set) var subscriptions: [Subscription] = []
    private(set) var signalsBySub: [String: [Signal]] = [:]
    var seenSignalIds: Set<String> = []

    var loading: Bool = false
    var lastError: String?
    var toast: String?

    /// Tab index — own state lives here, not in TabView, so we can deep-link
    /// from a notification tap.
    var selectedTab: Tab = .watchers

    /// Latest live-agent run for the watchers screen (visualizer overlay).
    var liveAgentSubId: String?

    let api: ApiService

    init(api: ApiService = ApiService(baseURL: BackendURL.url)) {
        self.api = api
    }

    enum Tab: Int, Hashable, CaseIterable {
        case watchers, alerts, signals, account
    }

    func bootstrap() async {
        do {
            try await api.bootstrapDevice()
            await refresh()
        } catch {
            lastError = "bootstrap: \(error.localizedDescription)"
        }
    }

    func refresh() async {
        loading = true
        defer { loading = false }
        do {
            let subs = try await api.listSubscriptions()
            self.subscriptions = subs.sorted { $0.createdAt > $1.createdAt }
            // Parallelise signal fetches — N subs = N concurrent requests, not N sequential.
            let pairs: [(String, [Signal])] = try await withThrowingTaskGroup(of: (String, [Signal]).self) { group in
                for sub in subs {
                    let id = sub.id
                    group.addTask { [api] in
                        let sigs = try await api.listSignals(subscriptionId: id, limit: 30)
                        return (id, sigs)
                    }
                }
                var out: [(String, [Signal])] = []
                for try await pair in group { out.append(pair) }
                return out
            }
            for (id, sigs) in pairs {
                signalsBySub[id] = sigs
                for s in sigs { seenSignalIds.insert(s.id) }
            }
        } catch {
            lastError = "refresh: \(error.localizedDescription)"
        }
    }

    func create(query: String, type: SubscriptionKind, cadenceSeconds: Int) async {
        do {
            let s = try await api.createSubscription(query: query, type: type, cadenceSeconds: cadenceSeconds)
            subscriptions.insert(s, at: 0)
            signalsBySub[s.id] = []
            Haptics.success()
            toast = "Watcher added: \(s.query)"
            // Authoritative resync runs in background — UI already optimistic,
            // no reason to make the caller (and the toast) wait on a full refetch.
            Task { await refresh() }
            // Kick the agent immediately so the user sees results without
            // waiting for the cadence boundary. `run` drives the LiveAgentView
            // overlay, fetches signals, and pushes for any new ones.
            Task { await run(s) }
        } catch {
            Haptics.error()
            lastError = "create failed: \(error.localizedDescription)"
        }
    }

    func delete(_ sub: Subscription) async {
        do {
            try await api.deleteSubscription(sub.id)
            subscriptions.removeAll { $0.id == sub.id }
            signalsBySub[sub.id] = nil
            Haptics.warning()
        } catch {
            lastError = "delete: \(error.localizedDescription)"
        }
    }

    /// Trigger an agent run — backend executes search → extract → insert,
    /// returns count of new signals. Newly arrived signals get mock pushes.
    func run(_ sub: Subscription) async {
        liveAgentSubId = sub.id
        do {
            _ = try await api.runSubscription(sub.id)
            let sigs = try await api.listSignals(subscriptionId: sub.id, limit: 30)
            let newOnes = sigs.filter { !seenSignalIds.contains($0.id) }
            signalsBySub[sub.id] = sigs
            for s in newOnes {
                seenSignalIds.insert(s.id)
                if let err = await MockPush.shared.deliver(
                    title: sub.query,
                    body: s.title,
                    subscriptionId: sub.id,
                    signalId: s.id
                ) {
                    lastError = err
                }
            }
            // give the visualizer a beat before dismissing
            try? await Task.sleep(nanoseconds: 600_000_000)
            liveAgentSubId = nil
            if !newOnes.isEmpty { Haptics.success() } else { Haptics.tap() }
        } catch {
            liveAgentSubId = nil
            Haptics.error()
            lastError = "run: \(error.localizedDescription)"
        }
    }

    /// Onboarding-side helper: drop a freshly-created subscription into local
    /// state so the home screen shows it immediately when onboarding dismisses,
    /// without waiting for a refresh round-trip.
    func injectSubscription(_ s: Subscription) {
        if !subscriptions.contains(where: { $0.id == s.id }) {
            subscriptions.insert(s, at: 0)
            signalsBySub[s.id] = []
        }
    }

    func removeSubscription(id: String) {
        subscriptions.removeAll { $0.id == id }
        signalsBySub[id] = nil
    }

    // ─── Derived ───
    func signals(for subId: String) -> [Signal] { signalsBySub[subId] ?? [] }

    var allSignalsRecent: [(Signal, Subscription)] {
        var out: [(Signal, Subscription)] = []
        for sub in subscriptions {
            for s in (signalsBySub[sub.id] ?? []) { out.append((s, sub)) }
        }
        return out.sorted { $0.0.firstSeenAt > $1.0.firstSeenAt }
    }

    var resolvedSubscriptions: [Subscription] {
        subscriptions.filter { sub in
            (signalsBySub[sub.id] ?? []).contains { $0.isResolved }
        }
    }

    var activeSubscriptions: [Subscription] {
        let resolvedIds = Set(resolvedSubscriptions.map(\.id))
        return subscriptions.filter { !resolvedIds.contains($0.id) }
    }

    func confirmedDate(for sub: Subscription) -> Date? {
        (signalsBySub[sub.id] ?? [])
            .compactMap(\.occursAt)
            .filter { $0 > Date() }
            .min()
    }
}

File: composeApp/src/commonMain/kotlin/com/notify/anything/notify/ui/AppState.kt
package com.notify.anything.notify.ui

import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.notify.anything.notify.ApiClient
import com.notify.anything.notify.SubscriptionType
import com.notify.anything.notify.platform.Notifier
import com.notify.anything.notify.platform.Prefs
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.delay
import kotlinx.coroutines.launch

enum class Tab { WATCHERS, ALERTS, SIGNALS, ACCOUNT }

/**
 * Mirrors the iOS `AppState`. Single owner of API client + observable
 * subscription/signal lists. Compose `mutableStateOf` triggers recomposition
 * when state changes — equivalent to SwiftUI's `@Observable`.
 */
class AppState(
    val scope: CoroutineScope,
    private val prefs: Prefs,
    private val notifier: Notifier,
    val defaultBaseUrl: String,
) {
    var baseUrl: String by mutableStateOf(prefs.getBaseUrl() ?: defaultBaseUrl)
        private set

    private var api: ApiClient = ApiClient(baseUrl)

    var deviceId: String? by mutableStateOf(prefs.getDeviceId()?.also { api.deviceId = it })
        private set

    var subscriptions: List<Subscription> by mutableStateOf(emptyList())
        private set

    var signalsBySub: Map<String, List<Signal>> by mutableStateOf(emptyMap())
        private set

    private val seenSignalIds: MutableSet<String> = mutableSetOf()

    var loading: Boolean by mutableStateOf(false)
        private set

    var lastError: String? by mutableStateOf(null)
    var toast: String? by mutableStateOf(null)

    var selectedTab: Tab by mutableStateOf(Tab.WATCHERS)
    var liveAgentSubId: String? by mutableStateOf(null)
        private set

    var detailSubscriptionId: String? by mutableStateOf(null)
    var showAddSheet: Boolean by mutableStateOf(false)

    fun bootstrap() {
        scope.launch {
            try {
                if (api.deviceId == null) {
                    val id = api.registerDevice("android-mock-${randomToken()}")
                    deviceId = id
                    prefs.putDeviceId(id)
                }
                refresh()
            } catch (t: Throwable) {
                lastError = "bootstrap: ${t.message ?: t.toString()}"
            }
        }
    }

    fun refresh() {
        scope.launch {
            loading = true
            try {
                val subs = api.listSubscriptions().map { it.toDomain() }
                    .sortedByDescending { it.createdAt }
                subscriptions = subs
                val pairs = subs.map { sub ->
                    async {
                        sub.id to api.listSignals(sub.id, limit = 30).map { it.toDomain() }
                    }
                }.awaitAll()
                val newMap = pairs.toMap()
                for ((_, sigs) in newMap) for (s in sigs) seenSignalIds.add(s.id)
                signalsBySub = newMap
            } catch (t: Throwable) {
                lastError = "refresh: ${t.message ?: t.toString()}"
            } finally {
                loading = false
            }
        }
    }

    fun create(query: String, type: SubscriptionType, cadenceSeconds: Int) {
        scope.launch {
            try {
                val s = api.createSubscription(query, type, cadenceSeconds).toDomain()
                subscriptions = listOf(s) + subscriptions
                signalsBySub = signalsBySub + (s.id to emptyList())
                toast = "Watcher added: ${s.query}"
                refresh()
            } catch (t: Throwable) {
                lastError = "create: ${t.message ?: t.toString()}"
            }
        }
    }

    fun delete(sub: Subscription) {
        scope.launch {
            try {
                api.deleteSubscription(sub.id)
                subscriptions = subscriptions.filterNot { it.id == sub.id }
                signalsBySub = signalsBySub - sub.id
            } catch (t: Throwable) {
                lastError = "delete: ${t.message ?: t.toString()}"
            }
        }
    }

    fun run(sub: Subscription) {
        scope.launch {
            liveAgentSubId = sub.id
            try {
                api.runSubscription(sub.id)
                val sigs = api.listSignals(sub.id, limit = 30).map { it.toDomain() }
                val newOnes = sigs.filter { it.id !in seenSignalIds }
                signalsBySub = signalsBySub + (sub.id to sigs)
                for (s in newOnes) {
                    seenSignalIds.add(s.id)
                    notifier.deliver(
                        title = sub.query,
                        body = s.title,
                        subscriptionId = sub.id,
                        signalId = s.id,
                    )
                }
                delay(600)
            } catch (t: Throwable) {
                lastError = "run: ${t.message ?: t.toString()}"
            } finally {
                liveAgentSubId = null
            }
        }
    }

    fun switchBackend(url: String) {
        scope.launch {
            try {
                api.close()
            } catch (_: Throwable) {}
            baseUrl = url
            prefs.putBaseUrl(url)
            prefs.putDeviceId(null)
            api = ApiClient(url)
            deviceId = null
            subscriptions = emptyList()
            signalsBySub = emptyMap()
            seenSignalIds.clear()
            toast = "Backend → $url"
            bootstrap()
        }
    }

    fun forgetDevice() {
        scope.launch {
            prefs.putDeviceId(null)
            deviceId = null
            api.deviceId = null
            bootstrap()
        }
    }

    fun signals(subId: String): List<Signal> = signalsBySub[subId] ?: emptyList()

    fun confirmedDate(sub: Subscription): Instant? =
        signals(sub.id).mapNotNull { it.occursAt }
            .filter { it > kotlinx.datetime.Clock.System.now() }
            .minOrNull()

    val resolvedSubscriptions: List<Subscription>
        get() = subscriptions.filter { sub ->
            signals(sub.id).any { it.isResolved }
        }

    val activeSubscriptions: List<Subscription>
        get() {
            val resolved = resolvedSubscriptions.map { it.id }.toSet()
            return subscriptions.filterNot { it.id in resolved }
        }

    val allSignalsRecent: List<Pair<Signal, Subscription>>
        get() = subscriptions.flatMap { sub ->
            signals(sub.id).map { it to sub }
        }.sortedByDescending { it.first.firstSeenAt }
}

private fun randomToken(): String {
    val chars = "abcdef0123456789"
    return (1..16).map { chars.random() }.joinToString("")
}

// Re-expose for callers that don't import kotlinx.datetime
typealias Instant = kotlinx.datetime.Instant
